posts/2023-04-13-rust-by-examples.adoc

```diff
@@ -0,0 +1,726 @@
+        if lines[i].contains(pattern) {
+            *idx = i;
+            return;
+        }
+    }
+    *idx = usize::MAX;
+}
+
+// function with a return value, FP style solution
+// if there is no semicolon after the last line, it is considered a return value
+// ("expr" is the same as "return expr;")
+fn search_pattern_iter(pattern: &str, lines: &[&str]) -> usize {
+    lines
+        .iter()  // iterate over the elements,
+                 // just like .stream() in Java (:Iterator<&str>)
+        .enumerate() // extend each element with an index, as a tuple,
+                     // just like .zip in Scala (:Iterator<(usize, &str)>)
+        .find(|(_, &line)| line.contains(pattern)) // find the first element where the closure
+                                                   // returns true (:Option<(usize, &str)>)
+        .map_or(usize::MAX, |(idx, _)| idx) // take the index from the tuple, if found,
+                                            // set MAX_USIZE otherwise (:usize)
+}
+----
+
+The program entry-point is the main() function in the main.rs file:
+
+[source,rust]
+----
+fn main() {
+    let lines = ["abcde", "defgh", "ghijk"];
+    let pattern = "gh";
+
+    let mut idx: usize = usize::MAX;
+    search_pattern_for(pattern, &lines, &mut idx);
+    // println!() is a macro. Macros can have variable number of arguments,
+    // functions must have fixed number of arguments
+    println!("Matching line: {}", if idx < lines.len() {lines[idx]} else {"NOT FOUND"});
+
+    let idx = search_pattern_iter(pattern, &lines);
+    println!("Matching line: {}", if idx < lines.len() {lines[idx]} else {"NOT FOUND"});
+}
+----
+
+Crates and modules are used to modularise your Rust code. We are not discussing them in this blog, but it is good to know that per default functions are private to the module. If you want to call a function from another module, it needs to be defined public. This is the same for structures and enums as well:
+[source,rust]
+----
+pub struct MyStruct {...}
+pub enum MyEnum {...}
+pub fn my_func() {...}
+----
+
+=== Ownership
+
+Ownership is a set of rules that govern how a Rust program manages memory. If any of the rules is violated, the program won't compile:
+
+* Each value in Rust has an owner.
+* There can only be one owner at a time.
+* When the owner goes out of scope, the value will be dropped.
+
+This is not an issue for primitive types, because they are small, and they are copied as an argument or a return value. Passing on non-primitive types to a function will move their ownership to the function and this ownership is not returned. For example:
+[source,rust]
+----
+fn return_match(pattern: String, lines: Vec<&str>) -> Option<String> {
+    lines
+        .iter()
+        .find(|&line| line.contains(&pattern))
+        .map(|&line| line.to_string())  // map &str to a String instance
+}
+
+fn example_2() {
+    let lines = vec!["abcde", "defgh", "ghijk"];
+    let pattern = "gh".to_string();
```

Comment:
A note of why `to_string` is called here (and in other places in the post) would be useful imo.

Reply:
I rewrote this example to pass on pattern as &str instead of String. I wanted to show passing function attributes by value, but passing String does not make much sense here. Passing the Vec by value will be enough for demonstration